core/src/services/supabase/writer.rs

```diff
@@ -0,0 +1,119 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use bytes::Bytes;
+use http::StatusCode;
+
+use super::core::*;
+use super::error::parse_error;
+use crate::ops::OpWrite;
+use crate::raw::*;
+use crate::*;
+
+pub struct SupabaseWriter {
+    core: Arc<SupabaseCore>,
+
+    op: OpWrite,
+    path: String,
+
+    buffer: oio::VectorCursor,
+    buffer_size: usize,
+}
+
+impl SupabaseWriter {
+    pub fn new(core: Arc<SupabaseCore>, path: &str, op: OpWrite) -> Self {
+        SupabaseWriter {
+            core,
+            op,
+            path: path.to_string(),
+            buffer: oio::VectorCursor::new(),
+            buffer_size: 8 * 1024 * 1024,
+        }
+    }
+
+    pub async fn upload(&self, bytes: Bytes) -> Result<()> {
+        let mut req = self.core.supabase_upload_object_request(
+            &self.path,
+            Some(bytes.len()),
+            self.op.content_type(),
+            AsyncBody::Bytes(bytes),
+        )?;
+
+        self.core.sign(&mut req)?;
+
+        let resp = self.core.send(req).await?;
+
+        match resp.status() {
+            StatusCode::OK => {
+                resp.into_body().consume().await?;
+                Ok(())
+            }
+            _ => Err(parse_error(resp).await?),
+        }
+    }
+}
+
+#[async_trait]
+impl oio::Write for SupabaseWriter {
+    async fn write(&mut self, bs: Bytes) -> Result<()> {
```

Comment:
`supabase_upload_object_request` seems always overwrite the same path and can't be used as resumeable upload. Please re-check the logic here. Take `s3` as a reference.

Reply:
Currently `supabase_upload_object_request()` does overwrite the same path with the same byte. But Supabase does not have something like `upload id`.

I am not sure about the correct behaviors you are pointing out, could you elaborate more on this?

Reply:
We should check the content length.

Only write content if content length match, otherwise, return Unsupported instead.